File: go/browser/deepquery.go
package browser

import _ "embed"

// DeepQueryJS defines shadow-piercing querySelector helpers, prepended to any
// browser-eval that must locate a node the way the offline matcher does: across
// shadow boundaries.
//
// Why this exists: capture is shadow-aware — probe.js walks node.shadowRoot and
// stamps data-sightmap-id on live shadow nodes, and the extracted tree flattens
// shadow subtrees in as ordinary children. So offline matching (and sel-probe's
// offline count, coverage, component queries) matches ACROSS shadow boundaries.
// A live document.querySelector does NOT cross shadow roots, so every live
// lookup that re-finds a matched node (property extraction, interaction,
// bounds, the sel-probe live count) was silently shadow-blind and disagreed
// with the corpus. These helpers close that gap. See schema.md "Selector model
// & shadow DOM".
//
// The traversal order matches probe.js's procNode exactly — a node's light-DOM
// children, fully (each recursively, including its own shadow content), before
// that node's own shadow-DOM children — so "first match" agrees with the
// corpus instead of surfacing shadow-DOM matches out of order.
//
//	__smDeepQueryAll(root, sel) — every element matching sel within root's tree
//	  AND every shadow tree nested beneath it. root may be the document, a
//	  shadow root, or an element; for an element the matches are its descendants
//	  (mirroring Element.querySelectorAll) plus descendants inside its own or any
//	  nested shadow root.
//	__smDeepQuery(root, sel) — the first such element, or null.
//
// Compose it once into any other embedded .js source that calls
// __smDeepQuery/__smDeepQueryAll (see e.g. actions.go's actionsJS), rather
// than prepending it again at each eval call site. The helpers are function
// declarations, so the script's trailing expression stays the completion
// value.
//
//go:embed deepquery.js
var DeepQueryJS string

File: go/browser/deepquery.js
// Shadow-piercing querySelector helpers, prepended to any browser-eval that
// must locate a node the way the offline matcher does: across shadow
// boundaries, in the same order probe.js's procNode flattens them (a node's
// own light-DOM subtree, fully, before that node's shadow-DOM subtree). See
// deepquery.go for why this exists and schema.md's "Selector model & shadow
// DOM" for the flattening rule this mirrors.
//
// DOM access goes through cached prototype accessors, NOT instance properties
// (node.children / el.matches / el.shadowRoot). An <input name="children"> makes
// HTMLFormElement's named getter return that control for form.children instead
// of the HTMLCollection, so a naive `for (const c of node.children)` throws
// "children is not iterable" on that form and the blanket catch below then
// returns a TRUNCATED result — silently 0 for any match after the form. (JetBlue's
// booker form has a "children" passenger-count control; that one collision made
// sel-probe and click/fill under-report.) The cached prototype accessors bypass
// any such named-property shadowing entirely.
//
// The accessors are cached INSIDE the function (not at top level): this file is
// prepended to CDP evals that run repeatedly in one tab context, so it must
// declare no top-level lexical bindings (a second eval of a top-level `const`
// throws "already declared"). Function declarations are re-eval-safe; keep it
// that way. See deepquery.go.
function __smDeepQueryAll(root, sel) {
  const childNodesOf = Object.getOwnPropertyDescriptor(
    Node.prototype,
    "childNodes",
  ).get;
  const matchesFn = Element.prototype.matches;
  const shadowRootGetter = Object.getOwnPropertyDescriptor(
    Element.prototype,
    "shadowRoot",
  ).get;
  // shadowRoot's getter lives on Element.prototype, so it is only valid for
  // element nodes (a Document/ShadowRoot root has none).
  const shadowOf = (node) =>
    node.nodeType === 1 ? shadowRootGetter.call(node) : null;

  const out = [];
  function visit(node) {
    for (const child of childNodesOf.call(node)) {
      if (child.nodeType !== 1) continue; // ELEMENT_NODE only
      if (matchesFn.call(child, sel)) out.push(child);
      visit(child);
      const sr = shadowOf(child);
      if (sr) visit(sr);
    }
  }
  try {
    visit(root);
    const rootShadow = shadowOf(root);
    if (rootShadow) visit(rootShadow);
  } catch (e) {
    return out;
  }
  return out;
}

function __smDeepQuery(root, sel) {
  const matches = __smDeepQueryAll(root, sel);
  return matches.length ? matches[0] : null;
}
